### ISSUE TITLE
Normal custom recipes added via CT is not shown in JEI worktable category

### ISSUE THREAD

**Issue by wormzjl** [codetaylor/pyrotech#142]
Currently normal crafting recipes added by CT does not reflect on pyrotech worktable.
Consider doing the recipe mirroring later in the loading phase?

**Comment by codetaylor** [comment]
*The issue template is provided for a reason, please use it.*

Thanks for the report!

> Consider doing the recipe mirroring later in the loading phase?

Recipes are not created during any loading phase. Vanilla recipes are converted to worktable recipes on-demand and cached. This is done to reduce memory overhead.

I have attempted to reproduce this using the following script:
```java
recipes.addShaped(<minecraft:string>, [
    [<minecraft:dirt>, <minecraft:dirt>, <minecraft:dirt>],
    [<minecraft:dirt>, null, <minecraft:dirt>],
    [<minecraft:dirt>, <minecraft:dirt>, <minecraft:dirt>]]);
```

Which works fine in both worktables:
![image](https://user-images.githubusercontent.com/1240716/62425691-ca793e00-b690-11e9-9eb0-5cfb78da2bdc.png)

HOWEVER, the CrT-added recipe is missing in JEI for both of the worktables, but ALSO for the vanilla table.

Please provide more of the information that was asked for in the issue template.

---

#### Script

*[replace]*

If your issue involves a CraftTweaker script, please place a **LINK** to your script here.

**DO NOT** paste your script here. Please use a service such as pastebin, hastebin, or gist.

*[/replace]*

#### Affected Versions

Do *not* use `latest`; please supply accurate version numbers.

* Forge:
* CraftTweaker:
* JEI:
* Pyrotech:
* Athenaeum:

**Comment by wormzjl** [comment]
My bad, I didn't test the actual crafting recipe, they are just not showing in worktable category but actually works.

**Comment by codetaylor** [comment]
This issue is not resolved. Please provide the information requested.

**Comment by wormzjl** [comment]
It's just some basic stuff like:
```
recipes.remove(<minecraft:hopper>);
recipes.addShaped(<minecraft:hopper>, [[<betterwithmods:material:14>, null, <betterwithmods:material:14>], [<betterwithmods:material:14>, <ore:chestWood>, <betterwithmods:material:14>], [null, <betterwithmods:material:14>, null]]);
recipes.addShaped(<minecraft:hopper>, [[<ore:ingotSteel>, null, <ore:ingotSteel>], [<ore:ingotSteel>, <ore:chestWood>, <ore:ingotSteel>], [null, <ore:ingotSteel>, null]]);
```

**Comment by codetaylor** [comment]
Ok, thanks. Versions?

Forge:
CraftTweaker:
JEI:
Pyrotech:
Athenaeum:

**Comment by wormzjl** [comment]
Forge: 1.12.2-14.23.5.2838
CraftTweaker: CraftTweaker2-1.12-4.1.19
JEI: jei_1.12.2-4.15.0.287
Pyrotech: pyrotech-1.12.2-1.2.9
Athenaeum: athenaeum-1.12.2-1.17.3

**Comment by codetaylor** [comment]
Awesome, thanks! :)

I'd like to get to the bottom of why JEI is not displaying the recipes.

Does it display your CrT recipes under the vanilla table?

**Comment by wormzjl** [comment]
Yeah it works fine with vanilla one

**Comment by codetaylor** [comment]
Hmm... ok. Yeah, it isn't showing in JEI under vanilla table for me.

**Comment by wormzjl** [comment]
Well, I've never seen a case where normal recipes added are not shown in vanilla, dev environment derpiness?

**Comment by katubug** [comment]
Hi! I'm having this same issue, so I figured I'd add some more data.

I'm using - 
Forge: 2847
CraftTweaker: 4.1.20
JEI: 4.15.0.291
Pyrotech: 1.4.16
Athenaeum: 1.17.4

Here are some examples of scripts that do not show correctly as belonging to the Worktables: https://pastebin.com/Wavz23iQ

All removing of recipes seems to work fine. And the recipe that I tested (dirt pile > dirt clumps) does work properly, it just doesn't show in JEI. I assume it's the same for the other recipes.

Please let me know if you need any logs or further tests! I'm happy to help.

**Comment by katubug** [comment]
I took the liberty of doing some additional stuff regardless, since I didn't feel like working on anything else. xD

I created an isolated instance with only CrT, JEI, and Pyrotech/dependencies. All versions are the same except for JEI, which is 4.15.0.293. I created a new world, then exited the game as instructed and loaded back up with the initially generated recipes present.

Here is the full test script file I used: https://pastebin.com/R4kTzWq7
Here is my latest.log: https://pastebin.com/rtB707Tf/
Here is my crafttweaker.log: https://pastebin.com/QUhrXkw4
And here is an album of screenshots of what I'm seeing with the relevant recipes: https://imgur.com/a/geGLF8f

Would it be beneficial for me to turn on JEI's debug mode and test? I'm not sure if that just appends things to the latest.log or what.

**Comment by codetaylor** [comment]
Should be fixed in `1.4.17`

**Comment by codetaylor** [comment]
For future reference:
* CrT added vanilla recipes don't show in the vanilla crafting category in the Pt dev environment because the deobf CrT doesn't contain the CrT JEI plugin
* CrT recipes were not showing in the Pt table because there weren't any JEI recipe handlers registered to handle instances of the CrT recipes, `MCRecipeShapeless` and `MCRecipeShaped` -- they don't extend any of the vanilla recipe classes that had registered handlers
